### ISSUE TITLE
CircleCI -> GitLab or GitHub CI

### ISSUE THREAD

**Issue by nedtwigg** [diffplug/spotless#950]
I switched us from Travis to CircleCI for two reasons:

1. "manual" pipelines so that our committer team can press a button to deploy, so that PR authors get their stuff delivered quickly (GitLab has this also)
2. windows runners

But I'm pissed off at it for these reasons:

1. it doesn't build [prospective merges](https://discuss.circleci.com/t/ability-to-checkout-merge-commits-not-just-branch-commits-maybe-even-make-it-the-default/28777) which has caused us [some problems](https://github.com/diffplug/spotless/pull/940#issuecomment-923222756)
2. it has randomly stopped building PRs, and we have a bunch open. I double-checked the dashboard and can find no reason why it stopped building PRs.
3. Our CI configuration has not changed, but all our CI pipelines just started failing with `Directory (/home/circleci/project) you are trying to checkout to is not empty and not a git repository.`
  - full green at a09a7b431296f39a49b447ecd39c61bc662778ad https://app.circleci.com/pipelines/github/diffplug/spotless/1249/workflows/61d6a1c5-7644-49b6-b09f-ae241be1eb92
  - one commit later its all red for no reason I can discern 85344bdbe94bac86a0723910dc6d32979d566bcc https://app.circleci.com/pipelines/github/diffplug/spotless/1250/workflows/8ff9cb65-895a-4c1b-9fcd-c9dc31bb1c94/jobs/7107

The point of paying for CI is to take things off my plate, but Circle keeps putting things back on it. I'm frustrated. I can't merge and release people's PRs if the CI is down, and I don't have time to debug CI that breaks without any changes on our side.

Apologies to PR authors, releases are gonna be frozen for a bit while I investigate a transition to GitLab.

**Comment by jbduncan** [comment]
No worries, @nedtwigg. That sounds awfully frustrating. If there's anything we can do to help, please let us know.

**Comment by joschi** [comment]
@nedtwigg Would using GitHub Actions be an alternative, too?

I've read https://github.com/diffplug/spotless/pull/570#issuecomment-624202702 and I understand the argument about a potential lock-in to GitHub.

This being said, I think the level of integration and the available actions could be a time saver compared to migrating the project to GitLab or integrating it to a degree with GitLab to be able to use GitLab CI.

**Comment by nedtwigg** [comment]
Well, I flushed our CircleCI project cache like so https://github.com/diffplug/spotless/commit/10b1834760bf4e794f1f5b9a2ae188130c5cabfe

We then started getting a new error related to the JRE 11. Looks like the CircleCI JRE 11 docker image changed, and one of our build plugins couldn't parse the java version anymore, which I fixed with https://github.com/diffplug/spotless/commit/1a9d7105982588fd2366192eb9cef622a647cbc4. and now our CI is working again.

While working on this, it looks like CircleCI lost some GitHub permissions somehow, and hopefully by relinking it our PRs will start working again.

If anybody submits a PR for a full or partial migration to GitHub actions, I would gratefully merge it. I tried to turn on GitLab CI, but the GitLab import got wedged somehow, and I'm frankly red hot with how much time this took and all these systems being so flaky.

**The point of running tests on tagged docker images is to make tests reproducible, so why-the-F does Circle silently change them underneath you**. That is insane. If the docker image needs bugfixes, it should submit a PR against the `config.yml` to bump the version number of the docker image.
